stdlib/Logging/src/ConsoleLogger.jl

```diff
@@ -0,0 +1,112 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+"""
+    ConsoleLogger(stream=STDERR, min_level=Info; min_level_prefix=Logging.Warn, show_limited=true)
+
+Logger with formatting optimized for readability in a text console, for example
+interactive work with the Julia REPL.
+
+Log levels less than `min_level` are filtered out.
+
+Message formatting can be controlled by setting keyword arguments.
+`min_level_prefix` controls whether the name of the log level will prefix the
+message (for example `@warn "blah"` being formatted as "Warning: blah").  The
+printing of large data structures in key value pairs is limited to what can be
+reasonably shown on the display when `show_limited=true`, by setting the
+`:limit` `IOContext` key during formatting.
+"""
+struct ConsoleLogger <: AbstractLogger
+    stream::IO
+    min_level::LogLevel
+    min_level_prefix::LogLevel
+    show_limited::Bool
+    message_limits::Dict{Any,Int}
+end
+function ConsoleLogger(stream::IO=STDERR, min_level=Info;
+                       min_level_prefix=Logging.Warn, show_limited=true)
+    ConsoleLogger(stream, min_level, min_level_prefix, show_limited, Dict{Any,Int}())
+end
+
+shouldlog(logger::ConsoleLogger, level, _module, group, id) =
+    get(logger.message_limits, id, 1) > 0
+
+min_enabled_level(logger::ConsoleLogger) = logger.min_level
+
+# Formatting of values in key value pairs
+showvalue(io, msg) = show(io, "text/plain", msg)
+function showvalue(io, e::Tuple{Exception,Any})
+    ex,bt = e
+    showerror(io, ex, bt; backtrace = bt!=nothing)
+end
+showvalue(io, ex::Exception) = showvalue(io, (ex,catch_backtrace()))
+
+function handle_message(logger::ConsoleLogger, level, message, _module, group, id,
+                        filepath, line; maxlog=nothing, kwargs...)
+    if maxlog != nothing && maxlog isa Integer
+        remaining = get!(logger.message_limits, id, maxlog)
+        logger.message_limits[id] = remaining - 1
+        remaining > 0 || return
+    end
+    levelstr = string(level)
+    level != Warn || (levelstr = "Warning")
+    color = level < Info  ? Base.debug_color() :
+            level < Warn  ? Base.info_color() :
+            level < Error ? Base.warn_color() :
+                            Base.error_color()
+    buf = IOBuffer()
+    iob = IOContext(buf, logger.stream)
+    if logger.show_limited
+        iob = IOContext(iob, :limit=>true)
+    end
+    msglines = split(chomp(string(message)), '\n')
+    prefixwithlevel = level >= logger.min_level_prefix
+    dsize = displaysize(logger.stream)
+    width = dsize[2]
+    locationstr = string(levelstr, " @ ", _module, " ", basename(filepath), ":", line)
+    singlelinewidth = 2 + length(msglines[1]) +
```

Comment:
Is it guaranteed that `msglines` has at least one element? (`split("")` gives an empty array in 0.6)